### DISCUSSION TITLE
Generate Agoric address from BIP-39 mnemonic seed

### DISCUSSION THREAD

**Discussion by thangld322** [Agoric/agoric-sdk#6994] [Q&A] [answered]
How can I convert a 24-word BIP-39 mnemonic seed to an Agoric address with programing?
I know I can get the address by importing the seed to an Agoric wallet like Keplr. But how can I do it by coding?

**Comment by dckc** [comment, marked as answer]
briefly... most of the answer is in https://github.com/Agoric/agoric-sdk/discussions/5830
but rather than `DirectSecp256k1Wallet.fromKey()` use [new DirectSecp256k1HdWallet(mnemonic: EnglishMnemonic, ...)](https://cosmos.github.io/cosmjs/latest/proto-signing/classes/DirectSecp256k1HdWallet.html#constructor)

**Reply by thangld322** [reply]
Excellent! But the output is still not the same as Keplr.
My code:
```javascript
import {DirectSecp256k1HdWallet} from '@cosmjs/proto-signing';

const makeAddr = async () => {
    const mnemonic = "account album blossom camp casino cheese cram deny draft extend gift grid helmet next outside pear praise property seed roast timber song wine tonight";
    const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {prefix:"agoric"});
    const accounts = await wallet.getAccounts();
    console.log({ accounts });
};

makeAddr().catch(err => console.error(err));
```
Output is `agoric16zf7grtfehwaftlpkhh87dpku8waqj6r4lrumd`
But if I import this mnemonic to Keplr, the address would be `agoric1s2fth0pryvhjsasrrwwrwm8g3nfaqjvc679fem`
If I change the prefix to `cosmos`, the output would be the same as Keplr (cosmos16zf7grtfehwaftlpkhh87dpku8waqj6r8zprtm).
What more do I need to do to make the output of Agoric the same as Keplr? 
Thank a lot!

**Reply by dckc** [reply]
Perhaps add `coinType: 564` in options?

```js
    bip44: {
      coinType: 564,
    },
    bech32Config: Bech32Address.defaultBech32Config("agoric"),
```

https://github.com/chainapsis/keplr-wallet/blob/c75e0b2d721879dcaaba8035fa43c00934be2d8b/packages/extension/src/config.ts#L1234-L1237

**Reply by thangld322** [reply]
```javascript
const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {prefix:"agoric", coinType: 564});
```
Looks like coinType is not an option for this function. The output is still the same.

**Reply by dckc** [reply]
oh... `coinType` is part of `hdPaths`. I have this code lying around...

```js
import { stringToPath } from '@cosmjs/crypto';

const hdPath = (coinType = 118, account = 0) =>
  stringToPath(`m/44'/${coinType}'/${account}'/0/0`);
```


see also: https://github.com/cosmos/cosmjs/blob/33271bc51c/packages/proto-signing/src/directsecp256k1hdwallet.ts#L116

**Reply by thangld322** [reply]
Works like a champ! Thank you so much!

**Reply by dckc** [reply]
Nice work assembling several clues into a working solution!

If you could share the whole thing assembled, that would be great.

**Comment by dckc** [comment]
for reference, I see this was also asked in https://community.agoric.com/t/convert-bip-39-mnemonic-seed-to-agoric-address/190/1
